src/os/bluestore/BlueStore_DB_Hash.cc

```diff
@@ -0,0 +1,678 @@
+    return db->get(cf, prefix, key, value);
+  }
+  int get(const string &prefix,
+          const char *key, size_t keylen,
+          bufferlist *value) override {
+    KeyValueDB::ColumnFamilyHandle cf = get_db_shard(prefix, key, keylen);
+    std::string s_key(key, keylen);
+    return db->get(cf, prefix, s_key, value);
+  }
+
+  int get(ColumnFamilyHandle cf_handle,
+                  const std::string &prefix,
+                  const std::set<std::string> &keys,
+                  std::map<std::string, bufferlist> *out) override {
+    ceph_assert(false && "invalid call");
+    return 0;
+  }
+
+  int get(ColumnFamilyHandle cf_handle,///< [in] Column family handle
+                  const std::string &prefix,    ///< [in] prefix or CF name
+                  const std::string &key,       ///< [in] key
+                  bufferlist *value) override {          ///< [out] value
+    ceph_assert(false && "invalid call");
+    return 0;
+  }
+
+  class WholeSpaceIteratorMerged_Impl: public WholeSpaceIteratorImpl {
+  private:
+    BlueStore_DB_Hash &db_hash;
+    ActiveShards::iterator shards_it;
+    /** for currently processed prefix, contains iterators to shards */
+    std::vector<KeyValueDB::Iterator> current_shards_iterators;
+    /** for currently processed prefix, marks first iterator that is not exhaused */
+    ssize_t position;
+    /*
+     * simulation of next/prev work
+     * v_______________________v   ---
+     * it0 it1 it2 it3 it4 it5 it6 it7
+     * next
+     * it1 it2 it0 it3 it4 it5 it6 it7
+     * next
+     * it2 it0 it3 it4 it1 it5 it6 it7
+     * next
+     * it2 it0 it3 it4 it1 it5 it6 it7
+     * next, it2 expired
+     * --- v___________________v   ---
+     * it2 it0 it3 it4 it1 it5 it6 it7
+     * next, it0 expired
+     * --- --- v_______________v   ---
+     * prev, requires checking before, revived it0
+     * --- v___________________v   ---
+     * it2 it0 it3 it4 it1 it5 it6 it7
+     * prev, checks before, but not revives
+     * it2 it0 it3 it4 it1 it5 it6 it7
+    */
+    bool open_shards() {
+      current_shards_iterators.clear();
+      for (auto& it: shards_it->second) {
+        Iterator wsi = db_hash.db->get_iterator_cf(it, shards_it->first);
+        wsi->seek_to_first();
+        if (wsi->valid())
+          current_shards_iterators.emplace_back(wsi);
+      }
+      position = 0;
+      std::sort(current_shards_iterators.begin(), current_shards_iterators.end(), KeyLess(*this));
+      return true;
+    }
+    int compare(const std::string& a, const std::string& b) {
+      rocksdb::Slice _a(a.data(), a.size());
+      rocksdb::Slice _b(b.data(), b.size());
+      return db_hash.comparator->Compare(_a, _b);
+    }
+    struct KeyLess {
+      WholeSpaceIteratorMerged_Impl& iter;
+      KeyLess(WholeSpaceIteratorMerged_Impl& iter) : iter(iter) {};
+      bool operator()(KeyValueDB::Iterator a, KeyValueDB::Iterator b) const
+      {
+        if (!a->valid())
```

Comment:
why isn't it necessary to look at ``b->valid()`` before calling ``b->key()`` below?